src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+#include <errno.h>
+#include <err.h>
+
+#define N_ELEMENTS(arr) (sizeof (arr) / sizeof ((arr)[0]))
+#define SD_LISTEN_FDS_START 3
+#define MAX_COCKPIT_WS_ARGS 5
+
+static struct instance_type
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
+
+static void
+handle_cockpit_ws_exited (int sigid)
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+              return;
```

Comment:
This may need a `break` instead. It can happen (in general, maybe not with this setup) that multiple processes terminate simultaneously, and there's only a single SIGCHLD. Even if there are reasons why this cannot happen in our specific situation, it just seems more robust to continue with the `while(waitpid)` loop.